### PR TITLE
Fix compile_commands.json for non-Windows

### PR DESCRIPTION
**Pull request by adiaaida**
On non-Windows platforms, we use CMAKE_EXPORT_COMPILE_COMMANDS, which
creates commands for every file and every project. We are only
interested in jit files, and we need each project to be separated so
that clang-tidy does not attempt to run on each file multiple times. If
it does, it replaces NULL with nullptr several times and we end up with
nullptrptrptr all over the code. This change takes the compile_commands
database generated by CMake, parses it and writes a new compile_commands
database with only the commands we are currently interested in.

### CODE REVIEW

**Comment by adiaaida** [comment]
@pgavlin @russellhadley @RussKeldorph PTAL
/cc: @dotnet/jit-contrib

File: src/jit-format/jit-format.cs [hunk 1]
```diff
@@ -563,6 +568,54 @@ public static void ParseBuildLog(string buildLog, string compileCommands, string
             }
         }
 
+        public static string rewriteCompileCommands (string compileCommandFile, string project)
+        {
+            string[] compileCommands = File.ReadAllLines(compileCommandFile);
```
Comment (RussKeldorph) [inline src/jit-format/jit-format.cs]: I can see that only relatively primitive parsing is required at this point, but why not parse as JSON?

File: src/jit-format/jit-format.cs [hunk 2]
```diff
@@ -563,6 +568,54 @@ public static void ParseBuildLog(string buildLog, string compileCommands, string
             }
         }
 
+        public static string rewriteCompileCommands (string compileCommandFile, string project)
```
Comment (RussKeldorph) [inline src/jit-format/jit-format.cs]: Please add a comment header describing the transformation this method intends to make.

File: src/jit-format/jit-format.cs (hunk 1 above)
Comment (adiaaida) [inline src/jit-format/jit-format.cs]: I was having a hard time finding documentation to help me figure out the errors I was seeing when I tried to parse it as JSON, so I went with this.

Comment (adiaaida) [inline src/jit-format/jit-format.cs]: I think I figured out how to parse the JSON. I'll update this in a few minutes

**Comment by adiaaida** [comment]
Updated the parser to use a Json parser.

PTAL.

File: src/jit-format/jit-format.cs [hunk 3]
```diff
@@ -563,6 +582,36 @@ public static void ParseBuildLog(string buildLog, string compileCommands, string
...
+                    newCommands.Add(new CompileCommand(command["directory"].Value<string>(),
+                                command["command"].Value<string>(), command["file"].Value<string>()));
+                }
+
+            }
+
+            // write commands back to a file.
+            string newCompileCommandsFileName = Path.Combine(Path.GetDirectoryName(compileCommandFile), "compile_commands_" + project + ".json");
+            string json = JsonConvert.SerializeObject(newCommands.ToArray(), Formatting.Indented);
+            System.IO.File.WriteAllText(newCompileCommandsFileName, json);
```
Comment (pgavlin) [inline src/jit-format/jit-format.cs]: Thank you for not overwriting the existing compile commands file :)

Comment (adiaaida) [inline src/jit-format/jit-format.cs]: You betcha.

File: src/jit-format/jit-format.cs [hunk 4]
```diff
@@ -343,6 +343,20 @@ public T ExtractDefault<T>(string name, out bool found)
         public string SourceDirectory { get { return _srcDirectory; } }
     }
 
+    public class CompileCommand
```
Comment (RussKeldorph) [inline src/jit-format/jit-format.cs]: Does this need to be public?

File: src/jit-format/jit-format.cs [hunk 5]
```diff
@@ -644,11 +693,11 @@ public static bool RunClangFormat(List<string> filenames, bool fix, bool verbose
 
                 if (verbose)
                 {
-                    Console.WriteLine("Running: clang-format {0} {1}", formatFix, filename);
+                    Console.WriteLine("Running: clang-format {0} -style=file {1}", formatFix, filename);
```
Comment (RussKeldorph) [inline src/jit-format/jit-format.cs]: Is this fix related?  If not, it should either be separated into its own change or accounted for in the description.

File: src/jit-format/jit-format.cs [hunk 6]
```diff
@@ -563,6 +582,36 @@ public static void ParseBuildLog(string buildLog, string compileCommands, string
...
+
+            foreach (JObject command in commands.Children<JObject>())
+            {
+                // Search for directory entries containing jit/<project>
+                if (command["directory"].Value<string>().Contains("jit/" + project))
+                {
+                    // Add the command to our list of new commands
+                    newCommands.Add(new CompileCommand(command["directory"].Value<string>(),
+                                command["command"].Value<string>().Replace("-I", "-isystem"), command["file"].Value<string>()));
+                }
```
Comment (RussKeldorph) [inline src/jit-format/jit-format.cs]: Use local temps to break up line and make it easier to read.

**Comment by RussKeldorph** [comment]
:shipit:

File: src/jit-format/jit-format.cs (hunk 5 above)
Comment (adiaaida) [inline src/jit-format/jit-format.cs]: No, it isn't. I'm not 100% sure it's necessary, though @pgavlin pointed it out on Friday. I've tried clang-format with and without -style=file, and I get the same results, so it's just a bit of paranoia

**Comment by adiaaida** [comment]
Apparently, clang-tidy doesn't actually like it when you give it a compile commands database that isn't called compile_commands.json. So I am going to have to change up the logic a little bit.

**Comment by adiaaida** [comment]
I have updated this to do the following to workaround the strange clang-tidy behavior I was seeing:

1. Move the compile_commands.json to compile_commands_full.json.
2. Parse compile_commands_full, and save away the results to compile_commands.json
3. Run clang-tidy.
4. Delete compile_commands.json and then move compile_commands_full back to compile_commands.json.

**Comment by adiaaida** [comment]
PTAL at the new changes.

File: src/jit-format/jit-format.cs (hunk 5 above)
Comment (RussKeldorph) [inline src/jit-format/jit-format.cs]: Ok, please update the description.

**Comment by RussKeldorph** [comment]
LGTM
